### ISSUE TITLE
raise version for @babel/core?

### ISSUE THREAD

**Issue by rhoehener** [expo/expo-cli#3570]
I'm trying to use Expo with MobX with annotations for state management, but keep running into issues:

https://github.com/babel/babel/issues/13452
https://github.com/mobxjs/mobx/issues/2981
https://github.com/mobxjs/mobx/discussions/2970

docs i'm trying to follow:
https://mobx.js.org/enabling-decorators.html
https://babeljs.io/docs/en/babel-plugin-proposal-decorators

**Comment by brentvatne** [comment]
we enable decorators by default if the `@babel/plugin-proposal-decorators` library is installed

https://github.com/expo/expo/blob/master/packages/babel-preset-expo/index.js#L63

**Comment by rhoehener** [comment]
But the babel docs say:

![image](https://user-images.githubusercontent.com/13407181/121592155-9c0ddc00-ca3a-11eb-81c1-8e75c9760155.png)

Which I cannot do because of babel/babel#13452

**Comment by brentvatne** [comment]
use yarn resolutions to override @babel/core if you want to update it to another version :)

**Comment by brentvatne** [comment]
i think something we can do here is change `@expo/metro-config` version to ^7.9.0 potentially. the problem is that we have encountered issues before when we track the latest versions of babel, whereas if we lock to a specific version then users will not have to deal with potentially out-of-nowhere accidental breaking changes that happen in patch releases

**Comment by brentvatne** [comment]
here's a working example of decorators from a brand new project, without any changes made outside of App.js:

```js
import { StatusBar } from 'expo-status-bar';
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

function myDecorator(target, name, descriptor) {
  return target;
}

@myDecorator
class App extends React.Component {
  render() {
    return (
      <View style={styles.container}>
        <Text>Open up App.js to start working on your app!</Text>
        <StatusBar style="auto" />
      </View>
    );
  }
}

export default App;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

**Comment by rhoehener** [comment]
Thanks, I have not used yarn so far. I will have another look tomorrow.

**Comment by brentvatne** [comment]
it looks like you can do similar things with npm with some third party package: https://stackoverflow.com/questions/52416312/npm-equivalent-of-yarn-resolutions
maybe it'll be built into npm soon: https://github.com/npm/rfcs/pull/129

**Comment by rhoehener** [comment]
I agree that Expo should not track the latest Babel version. Actually I appreciate the fact that Expo tries to provide version stability at least within SDK releases.

So I made another test:

- expo init (blank typescript)
- expo install mobx-react
- didn't change `babel.config.js` at all, based on your statement that Expo already supports decorators out of the box (meaning I ignore all instructions on MobX and Babel websites)
- added "experimentalDecorators": true to `tsconfig.json` (because Webstorm complained about this)

And indeed, my MobX basic test works, and the log statements appear in the expo terminal as well (both when clicking and automatically from the background task.

So I am going to give the MobX annotations another chance. 

Thanks again for your support!

```
import React from 'react';
import {Button, View, Text} from 'react-native';
import {observer} from "mobx-react";
import {makeObservable, observable, action} from "mobx";

@observer
export default class App extends React.Component<any, any> {
  @observable private _clickValue = 1;
  @observable private _autoValue = 1;

  constructor(props: any) {
    super(props);
    console.log("==================================");
    console.log("App init");
    console.log("==================================");
    makeObservable(this);
  }

  componentDidMount() {
    this.backgroundTask();
  }

  render() {
    return (
      <View style={{padding: 100}}>
        <Button title="Click" onPress={e => this.clickIncrement()}/>
        <Text>Click value: {this._clickValue}</Text>
        <Text>Auto value: {this._autoValue}</Text>
      </View>
    );
  }

  @action
  private clickIncrement() {
    this._clickValue++;
    console.log("clickIncrement: " + this._clickValue);
  }

  @action
  private autoIncrement() {
    this._autoValue++;
    console.log("autoIncrement: " + this._autoValue);
  }

  private async backgroundTask() {
    for (let i = 0; i < 100; i++) {
      this.autoIncrement();
      await this.sleep(2000);
    }
  }

  private async sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
```

**Comment by JLHwung** [comment]
> we have encountered issues before when we track the latest versions of babel, whereas if we lock to a specific version then users will not have to deal with potentially out-of-nowhere accidental breaking changes that happen in patch releases

Can you forward these issues to the Babel repo? 🙏 I can investigate what is going wrong and see if I can help.

`@babel/core` 7.13 provides an [`assumptions`](https://babel.dev/docs/en/assumptions) option that can optimize the output code size, and we will add more in the upcoming 7.15. Maybe worth give it a try and see if expo works with Babel 7.13?

**Comment by mogarick** [comment]
In my case I'm following mobx advice about avoiding annotations and using `makeObservable` but in order to track class properties [they need to be initialized](https://mobx.js.org/observable-state.html#limitations) on declaration or made [some babel configs](https://mobx.js.org/installation.html#use-spec-compliant-transpilation-for-class-properties) that are indeed only supported from version 7.13. Current expo-cli is 7.9 so I'm getting the assumptions error.

I tried to override the expo babel version by adding `resolutions` for my root project but it doesn't apply such overrides to the global `expo/cli` that builds and bundles the project when I run `expo start`. I can see the version via a console.log I added on my `babel.config.js`.

So how could I made expo-cli global to use babel/core 7.13+ in order it supports `assumptions`?

**Comment by brentvatne** [comment]
expo-cli global doesn't provide the babel version, it's local to your project.

**Comment by JLHwung** [comment]
This issue has somehow been addressed by https://github.com/expo/expo-cli/pull/3693. Expo now uses surcase so it won't provide the babel version like it [did](https://github.com/expo/expo-cli/blob/fcbe8a7dcc955add3817cfef1f1d92fbaa413ad8/packages/config/package.json#L35) before.

**Comment by mogarick** [comment]
> expo-cli global doesn't provide the babel version, it's local to your project.
> 
> the preset used is [babel-preset-expo](https://github.com/expo/expo/blob/master/packages/babel-preset-expo/package.json) which extends [metro-react-native-babel-preset](https://github.com/facebook/metro/blob/master/packages/metro-react-native-babel-preset/)

@brentvatne the weird thing is in my local project I have version `7.14.5` as dev dependency but still when I execute `expo start` my `babel.config.js` indicates that it's using `7.9.0`. Is there a way to trace the path of this babel executable?

I went to my project `nom_modlues/.bin` and the `babel` command there is a file that point to `../lib/babel` but I'm not sure what's the base dir this path is relative to because there is no `node_modules/lib` dir with a babel executable.

I check my globaly installed node modules but I can't find version `7.9.0` neither.
```
npm list -g --depth 1 |grep babel
│ ├── babel-runtime@6.26.0
│ ├── @babel/cli@7.14.8
│ ├── @babel/core@7.14.8
│ ├── @babel/plugin-proposal-class-properties@7.14.5
│ ├── @babel/plugin-proposal-decorators@7.14.5
│ ├── @babel/preset-env@7.14.8
│ ├── @babel/preset-typescript@7.14.5
│ ├── babel-loader@8.2.2
│ ├── babel-plugin-module-resolver@4.1.0
│ ├── babel-plugin-root-import@6.6.0
│ ├── babel-plugin-transform-export-extensions@6.22.0
│ ├── babel-preset-expo@8.3.0
│ ├── metro-react-native-babel-preset@0.62.0
│ ├── @babel/cli@7.14.5
│ ├── @babel/core@7.14.6
│ ├── @babel/plugin-proposal-class-properties@7.14.5
│ ├── @babel/plugin-proposal-decorators@7.14.5
│ ├── @babel/preset-env@7.14.7
│ ├── @babel/preset-typescript@7.14.5
│ ├── babel-loader@8.2.2
│ ├── babel-plugin-module-resolver@4.1.0
│ ├── babel-plugin-root-import@6.6.0
│ ├── babel-plugin-transform-export-extensions@6.22.0
│ ├── babel-preset-expo@8.3.0
│ ├── metro-react-native-babel-preset@0.62.0
```

any hints or help you could provide would be very appreciated.

Thank you.

**Comment by mogarick** [comment]
@brentvatne Checking again the expo error I can see that it is indeed using `7.9.0` from its own `node_modules` so I'm confused with your last comment about expo-cli using the local project babel. Can you help me to understand please?

```
› Open in the web browser...
Configuring Babel(ver. 7.9.0) Root Project...

The project's Babel config is invalid: Unknown option: .assumptions. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.
(node:7190) UnhandledPromiseRejectionWarning: Error: Unknown option: .assumptions. Check out https://babeljs.io/docs/en/babel-core/#options for more information about options.
    at throwUnknownError (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/validation/options.js:123:27)
    at /Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/validation/options.js:108:5
    at Array.forEach (<anonymous>)
    at validateNested (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/validation/options.js:84:21)
    at validate (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/validation/options.js:75:10)
    at /Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/config-chain.js:169:34
    at cachedFunction (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/caching.js:62:27)
    at cachedFunction.next (<anonymous>)
    at evaluateSync (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/gensync/index.js:251:28)
    at sync (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/gensync/index.js:89:14)
    at buildRootChain (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/config-chain.js:89:27)
    at buildRootChain.next (<anonymous>)
    at loadPrivatePartialConfig (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/partial.js:95:62)
    at loadPrivatePartialConfig.next (<anonymous>)
    at Function.<anonymous> (/Users/mogarick/n/lib/node_modules/expo-cli/node_modules/@babel/core/lib/config/partial.js:120:25)
    at Generator.next (<anonymous>)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:7190) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:7190) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by mogarick** [comment]
> This issue has somehow been addressed by #3693. Expo now uses surcase so it won't provide the babel version like it [did](https://github.com/expo/expo-cli/blob/fcbe8a7dcc955add3817cfef1f1d92fbaa413ad8/packages/config/package.json#L35) before.

@JLHwung it looks that's right. After upgrading my expo-cli to the latest it includes the sucrase change and now my project transpiling is done using my declared babel version.

Now my new problem is that by following mobx recommended configs it somehow messes up with amplify code and my app gets gets stuck when using async/await for amplify Auth methods. So it looks I still gonna have some fun.
